client/src/components/mapping/FhirMappingPanel/FhirResourceTree/index.tsx

```diff
@@ -84,137 +67,122 @@ const FhirResourceTree = ({ onClickCallback }: Props) => {
-    deleteNodeCallback?: Function
-  ): ITreeNode<Node> => {
-    const renderNodeLabel = () => {
-      const nodeLabel = (
-        <NodeLabel
-          name={path.parent?.isArray ? path.definition.name : path.tail()}
-          type={path.types.join(' | ')}
-          addNodeCallback={addNodeCallback}
-          deleteNodeCallback={deleteNodeCallback}
-        />
-      );
-
-      if (!path.definition.definition) return nodeLabel;
-      return (
-        <Tooltip boundary={'viewport'} content={path.definition.definition}>
-          {nodeLabel}
-        </Tooltip>
-      );
-    };
+  // To update secondary label on user input
+  const updateSecondaryLabel = (node: ITreeNode<Attribute>) => {
+    if (!node.nodeData) return;
+    node.secondaryLabel = secondaryLabel(node.nodeData, node.childNodes || []);
+    node.childNodes?.forEach(child => updateSecondaryLabel(child));
+  };
 
-    return {
+  const createNode = (
+    attribute: Attribute,
+    childNodes: ITreeNode<Attribute>[],
+    parentArray?: ITreeNode<Attribute>
+  ): ITreeNode<Attribute> => {
+    const node: ITreeNode<Attribute> = {
       childNodes: childNodes,
-      hasCaret: !path.isPrimitive || path.isArray,
-      icon: path.isArray
+      hasCaret: !attribute.isPrimitive || attribute.isArray,
+      icon: attribute.isArray
         ? 'multi-select'
-        : !path.isPrimitive
+        : !attribute.isPrimitive
         ? 'folder-open'
         : 'tag',
-      id: path.tail(),
-      label: renderNodeLabel(),
-      secondaryLabel: secondaryLabel(path, childNodes),
-      nodeData: path
+      id: attribute.tail,
+      secondaryLabel: secondaryLabel(attribute, childNodes),
+      nodeData: attribute,
+      label: ''
     };
+    node.label = (
+      <NodeLabel
+        attribute={attribute}
+        addNodeCallback={() => addNodeToArray(node)}
+        deleteNodeCallback={() => deleteNodeFromArray(node, parentArray!)}
+      />
+    );
+    return node;
   };
 
   const deleteNodeFromArray = (
-    stateNodes: ITreeNode<Node>[],
-    path: Node
-  ): ITreeNode<Node>[] => {
+    deletedNode: ITreeNode<Attribute>,
+    arrayNode: ITreeNode<Attribute>
+  ) => {
     // First, we delete all the corresponding attributes in DB
+    const deleted = deletedNode.nodeData!;
+    const array = arrayNode.nodeData!;
     deleteAttributes({
       variables: {
-        startsWith: path.serialize()
+        startsWith: deleted.path
       }
     });
     // Same in Redux store
-    dispatch(removeAttributesFromMap(path.serialize()));
-
-    // Then we find the parent from which we want to remove
-    // the child node.
-    const parent = findNode(stateNodes, path.parent!);
-    if (!parent || !parent.childNodes) return stateNodes;
+    dispatch(removeAttributesFromMap(deleted.path));
+    // and remove the item in the parent attribute as well
+    array.removeItem(deleted);
 
-    if (parent.childNodes.length === 1) {
+    if (arrayNode.childNodes?.length === 1) {
```

Comment:
Shouldn't you check that the length of the array is 0 now that you removed the element?

Reply:
I removed the attribute (ie: arrayNode.nodeData) but not the childNode